src/gov-action-contracts/AIPs/AIPArbOS30/parent_contracts/AIPArbOS30UpgradeChallengeManagerAction.sol

```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.16;
+
+import "@arbitrum/nitro-contracts/src/precompiles/ArbSys.sol";
+import "@arbitrum/nitro-contracts/src/osp/IOneStepProofEntry.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+import "@openzeppelin/contracts/proxy/transparent/ProxyAdmin.sol";
+import "../../../address-registries/L1AddressRegistry.sol";
+
+interface IChallengeManagerUpgradeInit {
+    // TODO
+    function postUpgradeInit(IOneStepProofEntry osp_, bytes32 condRoot, IOneStepProofEntry condOsp)
+        external;
+    function osp() external returns (address);
+}
+
+// @notice set wasm module root and upgrade challenge manager for stylus ArbOS upgrade
+contract AIPArbOS30UpgradeChallengeManagerAction {
+    L1AddressRegistry public immutable l1AddressRegistry;
+    bytes32 public immutable newWasmModuleRoot;
+    ProxyAdmin public immutable govProxyAdmin;
+    address public immutable newChallengeManagerImpl;
+    IOneStepProofEntry public immutable osp;
+    bytes32 immutable condRoot;
+    IOneStepProofEntry public immutable condOsp;
+
+    constructor(
+        L1AddressRegistry _l1AddressRegistry,
+        bytes32 _newWasmModuleRoot,
+        ProxyAdmin _govProxyAdmin,
+        address _newChallengeManagerImpl,
+        IOneStepProofEntry _osp,
+        bytes32 _condRoot,
+        IOneStepProofEntry _condOsp
```

Comment:
Instead of setting this in constructor, we can also have the upgrade action to use the current (pre-upgrade) wasm root and osp for `cond` during perform; not sure which is better

Reply:
since we're now planning on deploying two new OSP contracts I believe this no longer applies?